vizro-core/tests/unit/vizro/models/_components/test_graph.py

```diff
@@ -1,33 +1,90 @@
+        figure = vm.Graph(figure=px.bar(data_frame=gapminder, x="year", y="pop", title=title)).__call__()
+
+        assert figure.layout.margin.t == expected
+        assert figure.layout.template.layout.margin.t == 64
+        assert figure.layout.template.layout.margin.l == 80
+        assert figure.layout.template.layout.margin.b == 64
+        assert figure.layout.template.layout.margin.r == 12
+
+
+class TestProcessFigureDataFrame:
+    def test_process_figure_data_frame_str_df(self, standard_px_chart_with_str_dataframe, gapminder):
+        data_manager["gapminder"] = gapminder
+        graph_with_str_df = vm.Graph(
+            id="text_graph",
+            figure=standard_px_chart_with_str_dataframe,
         )
+        assert data_manager._get_component_data("text_graph").equals(gapminder)
+        assert graph_with_str_df.figure["data_frame"] == "gapminder"
 
+    def test_process_figure_data_frame_df(self, standard_px_chart, gapminder):
+        graph_with_df = vm.Graph(
+            id="text_graph",
+            figure=standard_px_chart,
+        )
+        assert data_manager._get_component_data("text_graph").equals(gapminder)
+        with pytest.raises(KeyError, match="'data_frame'"):
+            graph_with_df.figure["data_frame"]
 
-@pytest.mark.parametrize("title, expected", [(None, 24), ("Test", None)])
-def test_title_margin_adjustment(gapminder, title, expected):
-    figure = vm.Graph(figure=px.bar(data_frame=gapminder, x="year", y="pop", title=title)).__call__()
 
-    assert figure.layout.margin.t == expected
-    assert figure.layout.template.layout.margin.t == 64
-    assert figure.layout.template.layout.margin.l == 80
-    assert figure.layout.template.layout.margin.b == 64
-    assert figure.layout.template.layout.margin.r == 12
+class TestBuild:
+    def test_create_empty_fig(self, expected_empty_chart):
+        result = create_empty_fig("NO DATA")
+        assert result == expected_empty_chart
+
+    def test_graph_build(self):
+        graph = vm.Graph(
+            id="text_graph",
+            figure=px.scatter(
+                data_frame=px.data.gapminder(),
+                x="gdpPercap",
+                y="lifeExp",
+                size="pop",
+                color="continent",
+                hover_name="country",
+                size_max=60,
+            ),
+        )
+        result = graph.build()
+
+        assert result.color == "grey"
+        assert result.parent_className == "chart_container"
+        assert result.children.id == "text_graph"
+        assert result.children.className == "chart_container"
+        assert result.children.config == {
+            "autosizable": True,
+            "frameMargins": 0,
+            "responsive": True,
+        }
+        # result.children.figure.layout.template is ignored
+        assert isinstance(result, dcc.Loading)
+        assert isinstance(result.children.figure, go.Figure)
```

Comment:
Would remove all of this and instead do a json comparison of the final component. This will then cover all of your checks above and more. The reason why we opted for this is because in very nested objects you would have numerous `isinstance` checks and equality checks. The json comparison is a more efficient way to do so.

Would first create a fixture of the expected result:

```
@pytest_fixture()
def expected_graph():
     return dcc.Loading(
            dcc.Graph(
                id="text_graph",
                figure=create_empty_fig("NO DATA"),
                config={
                    "autosizable": True,
                    "frameMargins": 0,
                    "responsive": True,
                },
                className="chart_container",
            ),
            color="grey",
            parent_className="chart_container",
        )
```

```suggestion
result = json.loads(json.dumps(graph.build(), cls=plotly.utils.PlotlyJSONEncoder))
expected = json.loads(json.dumps(expected_graph, cls=plotly.utils.PlotlyJSONEncoder))
assert result == expected
```